src/Conversion/ONNXToKrnl/Tensor/Transpose.cpp

```diff
@@ -20,83 +20,192 @@ using namespace mlir;
 namespace onnx_mlir {
 
 struct ONNXTransposeOpLowering : public ConversionPattern {
+  using MDBuilder = MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl,
+      MemRefBuilder, MathBuilder>;
+
   ONNXTransposeOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
       : ConversionPattern(
             typeConverter, mlir::ONNXTransposeOp::getOperationName(), 1, ctx) {}
 
   LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
       ConversionPatternRewriter &rewriter) const final {
-    ONNXTransposeOpAdaptor operandAdaptor(operands);
-    ONNXTransposeOp transposeOp = llvm::cast<ONNXTransposeOp>(op);
     Location loc = op->getLoc();
-    MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl> create(
-        rewriter, loc);
+    MDBuilder create(rewriter, loc);
 
     // Operands and attributes.
+    ONNXTransposeOpAdaptor operandAdaptor(operands, op->getAttrDictionary());
     Value data = operandAdaptor.data();
-    auto permAttr = transposeOp.perm();
+    Optional<ArrayAttr> permAttr = operandAdaptor.perm();
 
+    // Input and output types.
     MemRefType inMemRefType = data.getType().cast<MemRefType>();
-    uint64_t inRank = inMemRefType.getShape().size();
-    // Convert the output type to MemRefType.
     Type outConvertedType =
         typeConverter->convertType(*op->result_type_begin());
     assert(outConvertedType && outConvertedType.isa<MemRefType>() &&
            "Failed to convert type to MemRefType");
     MemRefType outMemRefType = outConvertedType.cast<MemRefType>();
-    uint64_t outRank = outMemRefType.getShape().size();
 
     // Get shape.
     ONNXTransposeOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
     shapeHelper.computeShapeAndAssertOnFailure();
+    DimsExpr &outDims = shapeHelper.getOutputDims();
```

Comment:
I like the overall idea a lot, namely to try to move data in chunks. I wonder if we could not create an infrastructure that would allow this type of optimization to be used elsewhere too, with an interface that would be easy to use. 

krnlOptDataMove(loop indices, vector of LB for loop iteration, vector of UB for loop iteration, memref from, memref to, set of affine map access function from, set of affine map access function to)

if there is an easy way to check which affine function are identical (is there such a check?), and/or which are continuous, then we could automatically detect which ones can be combined.

Alternatively, we could also generate the normal code, and then have an op that gather contiguous loads. That could be run after memref normalization, so it would handle even cases with complicated maps. I would even check if there is not already something like this in MLIR.

I understand that this might be for another PR, just wanted to see if this approach can be applied to all the other memcpy that we experience when lowering ops.

Reply:
Yes, I agree that it would be great to generalize this for other situations.